### ISSUE TITLE
grpc-netty-shaded doesn't use TCP_USER_TIMEOUT on ARMv8

### ISSUE THREAD

**Issue by davidkilliansc** [grpc/grpc-java#7956]
Hi, I had been running w/ the native transport and noticed it had unexpectedly switched back to NIO at some point.  After some investigation I realized the switch happened when I moved to running on ARM.  Looks like grpc-netty-shaded doesn't include the ARM artifacts.  

When I switched to depending on grpc-netty rather than grpc-netty-shaded, that didn't work either at first.  When looking at profiles I found it was still using grpc-netty-shaded classes.   I found the shaded package was being brought into my dependency closure from some other dependencies I have.  I think what's happening is when I use ServerBuilder, it's discovering two NettyServerProviders on the classpath: one from grpc-netty and another from grpc-netty-shaded.   They're both the same priority and the shaded version ends up being chosen (which in turn won't work with ARM).    

To summarize IIUC there's two issues: grpc-netty-shaded doesn't included ARM, and there doesn't seem to be a way to avoid using the shaded classes as long as anything brings in the shaded dependency.   I'm going to workaround the latter by directly using NettyServerBuilder rather than ServerBuilder.

**Comment by sanjaypujare** [comment]
> ...
> To summarize IIUC there's two issues: grpc-netty-shaded doesn't included ARM, 

I suspect that won't be fixed soon...

> and there doesn't seem to be a way to avoid using the shaded classes as long as anything brings in the shaded dependency. I'm going to workaround the latter by directly using NettyServerBuilder rather than ServerBuilder.

What kind of fix do you expect? Should `NettyServerProvider` use reflection to decide what priority to return based on it's shaded/unshaded full name? Or something else?

**Comment by davidkilliansc** [comment]
> What kind of fix do you expect? Should NettyServerProvider use reflection to decide what priority to return based on it's shaded/unshaded full name? Or something else?

Non-shaded being higher priority would give me the behavior I want/expected, but unfortunately that might not be the right priority for others?  e.g. if I depended on grpc-netty-shaded but one of my transitive dependencies brought in grpc-netty, then prioritizing the non-shaded provider would be the same problem.   

If we at least preferred the provider that matched gRPC versions w/ the ServerBuilder I'm using, that could help.  When I depended on shaded, my gradle setup would auto-resolve conflicts w/ my dependency's shaded versions by forcing resolution to the version of shaded I directly depended on.  Now that I don't depend on shaded, the version conflict is gone, so an older version of shaded is in my classpath, and so not only am I inadvertently using the shaded dependency, but it's coming form an older build of grpc-java than the other grpc-java classes I'm using.    

Another problem is netty-related system properties I had set are no longer taking effect.  I set the non-shaded version of the system property, but the shaded version what's needed.  If I have both shaded and non-shaded in my classpath there's no way to reason about which system properties I should be setting.  

Only other thing I can think of is the user code somehow carries the intent w/ it of whether shaded should be used or not rather than it being dependent on what happens to be in the classpath.  For example, if there were different ServerBuilders, one in shaded and on in non-shaded, then which ServerBuilder I use determines which netty dependency I'm intending to use.

**Comment by sanjaypujare** [comment]
> ...
> Only other thing I can think of is the user code somehow carries the intent w/ it of whether shaded should be used or not rather than it being dependent on what happens to be in the classpath. For example, if there were different ServerBuilders, one in shaded and on in non-shaded, then which ServerBuilder I use determines which netty dependency I'm intending to use.

Can you propose something concrete about using different ServerBuilders that can be considered for potential implementation?

**Comment by ejona86** [comment]
Some things need to be clarified:
1. It seems grpc-netty-shaded _does_ support 32bit ARM, but doesn't use epoll for you. This might have reduced performance and might lose some socket options, but should mostly function fine
2. grpc-netty-shaded _does_ support epoll on ARM64

So I have to assume you are on ARMv7 (or similar) _or_ you are using an older version of gRPC. What version of gRPC are you using, what platform are you on, and what problem are you experiencing (e.g., error message)?

> there doesn't seem to be a way to avoid using the shaded classes as long as anything brings in the shaded dependency. I'm going to workaround the latter by directly using NettyServerBuilder rather than ServerBuilder.

The ServerBuilder API is the "I don't care what transport I use, as long as it works" API. Using NettyServerBuilder directly is appropriate if you want to choose a specific transport. I wouldn't call that a workaround, although I also don't see any clear misbehavior mentioned either.

> When I depended on shaded, my gradle setup would auto-resolve conflicts w/ my dependency's shaded versions by forcing resolution to the version of shaded I directly depended on.

If there is version tear across gRPC artifacts, it can cause lots of random problems. This doesn't seem to be related to grpc-netty-shaded specifically. You can use grpc-bom to _help_ get the versions in alignment. grpc-bom is helpful for _upgrading_ to a common version, but if any transitive dependency gets a newer version of grpc than you you'd need to bump the grpc-bom version.

**Comment by davidkilliansc** [comment]
> What version of gRPC are you using, what platform are you on, and what problem are you experiencing (e.g., error message)?

grpc-java version 1.35.0 on aarch64.  I don't see any native epoll artifact for aarch64 in META-INF/native/ when I crack open the grpc-netty-shaded v1.35.0 jar.  Same with the most recent 1.36.0 jar.  What am I missing? 

The problem I experienced was initially:  

> I had been running w/ the native transport and noticed it had unexpectedly switched back to NIO

I was profiling and tuning some of the Netty buffer-related system properties in a pre-production environment when I happened to notice the codepaths related to native Epoll were no longer triggering.  Our intention is that native Epoll is used for performance & cost reasons.   Concurrently, there was a separate stream of work going on that would utilize UNIX Domain Sockets, where we explicitly configured via `NettyServerBuilder` rather than `ServerBuilder`, but in that case ran into the following error: 

```
java.lang.UnsatisfiedLinkError: failed to load the required native library
	at io.grpc.netty.shaded.io.netty.channel.epoll.Epoll.ensureAvailability(Epoll.java:80)
	at io.grpc.netty.shaded.io.netty.channel.epoll.EpollEventLoop.<clinit>(EpollEventLoop.java:51)
	at io.grpc.netty.shaded.io.netty.channel.epoll.EpollEventLoopGroup.newChild(EpollEventLoopGroup.java:150)
	at io.grpc.netty.shaded.io.netty.channel.epoll.EpollEventLoopGroup.newChild(EpollEventLoopGroup.java:35)
	at io.grpc.netty.shaded.io.netty.util.concurrent.MultithreadEventExecutorGroup.<init>(MultithreadEventExecutorGroup.java:84)
	at io.grpc.netty.shaded.io.netty.util.concurrent.MultithreadEventExecutorGroup.<init>(MultithreadEventExecutorGroup.java:58)
	at io.grpc.netty.shaded.io.netty.util.concurrent.MultithreadEventExecutorGroup.<init>(MultithreadEventExecutorGroup.java:47)
	at io.grpc.netty.shaded.io.netty.channel.MultithreadEventLoopGroup.<init>(MultithreadEventLoopGroup.java:59)
	at io.grpc.netty.shaded.io.netty.channel.epoll.EpollEventLoopGroup.<init>(EpollEventLoopGroup.java:112)
	at io.grpc.netty.shaded.io.netty.channel.epoll.EpollEventLoopGroup.<init>(EpollEventLoopGroup.java:99)
	at io.grpc.netty.shaded.io.netty.channel.epoll.EpollEventLoopGroup.<init>(EpollEventLoopGroup.java:60)
	<snip>
Caused by: java.lang.UnsatisfiedLinkError: could not load a native library: io_grpc_netty_shaded_netty_transport_native_epoll_aarch_64
	at io.grpc.netty.shaded.io.netty.util.internal.NativeLibraryLoader.load(NativeLibraryLoader.java:226)
	at io.grpc.netty.shaded.io.netty.channel.epoll.Native.loadNativeLibrary(Native.java:250)
	at io.grpc.netty.shaded.io.netty.channel.epoll.Native.<clinit>(Native.java:69)
	at io.grpc.netty.shaded.io.netty.channel.epoll.Epoll.<clinit>(Epoll.java:39)
	at java.base/java.lang.Class.forName0(Native Method)
	at java.base/java.lang.Class.forName(Class.java:315)
	at io.grpc.netty.shaded.io.grpc.netty.Utils.isEpollAvailable(Utils.java:291)
	at io.grpc.netty.shaded.io.grpc.netty.Utils.<clinit>(Utils.java:109)
	at io.grpc.netty.shaded.io.grpc.netty.NettyServerBuilder.<clinit>(NettyServerBuilder.java:84)
	at io.grpc.netty.shaded.io.grpc.netty.NettyServerProvider.builderForPort(NettyServerProvider.java:40)
	at io.grpc.netty.shaded.io.grpc.netty.NettyServerProvider.builderForPort(NettyServerProvider.java:25)
	at io.grpc.ServerBuilder.forPort(ServerBuilder.java:41)
```

> The ServerBuilder API is the "I don't care what transport I use, as long as it works" API. Using NettyServerBuilder directly is appropriate if you want to choose a specific transport. I wouldn't call that a workaround, although I also don't see any clear misbehavior mentioned either.

When I saw that Netty supported native Epoll on ARM, and that I had a high enough version of gRPC to pick up a new enough version of Netty that'd include the support, then I expected that my dependency on on grpc-netty-shaded would result in me using native Epoll on ARM just like it does when on x86.   If I should assume less about transport implementation when using ServerBuilder that's fair.  Part of the reason I called it a workaround is NettyServerBuilder is marked experimental so it gives the impression I'm doing something a little unsavory in order to choose to pin the transport impl :)

**Comment by ejona86** [comment]
>  don't see any native epoll artifact for aarch64 in META-INF/native/ when I crack open the grpc-netty-shaded v1.35.0 jar.

Oh, I was seeing the tcnative file. Looks like the problem is there's no uber jar for epoll, so we [are only bringing in the x86_64 classifier](https://github.com/grpc/grpc-java/blob/f872c6059dfe3c28400db7c1e760a95385aecc14/build.gradle#L172). That seems fixable in netty/shaded but we'd want to see how the duplicate class files are handled if we bring in epoll twice.

> Our intention is that native Epoll is used for performance & cost reasons.

That seems the only relevant problem statement to grpc-netty-shaded. Given we've not done any performance testing armv8, nothing actually guarantees that the epoll transport is actually faster. I'll note that we didn't actually include epoll in grpc-netty-shaded for performance reasons but rather [for TCP_USER_TIMEOUT](https://github.com/grpc/grpc-java/blob/f872c6059dfe3c28400db7c1e760a95385aecc14/netty/src/main/java/io/grpc/netty/NettyClientTransport.java#L240).

I'll note that there are multiple possible performance optimizations you may choose to make by using NettyServerBuilder directly; using ServerBuilder doesn't not guarantee "optimum" performance for every possible situation.

That said, I do think we want to include the epoll aarch64 binary in grpc-netty-shaded. But we probably aren't going to rush in doing so.

> Concurrently, there was a separate stream of work going on that would utilize UNIX Domain Sockets, where we explicitly configured via NettyServerBuilder rather than ServerBuilder

This shouldn't be impacted by grpc-netty-shaded then. Did you depend on `io.netty:netty-transport-native-epoll:4.1.52.Final:linux-aarch64`?

> Part of the reason I called it a workaround is NettyServerBuilder is marked experimental so it gives the impression I'm doing something a little unsavory in order to choose to pin the transport impl :)

The problem is that the precise transports supported is not guaranteed in terms of API stability. For example, if we upgrade to a Netty 5 we will want to drop support for Netty 4.1 at some point. The configuration provided on the API can also be things we can't guarantee long-term as our implementation changes. That said, there's quite a few reasonable reasons to use NettyServerBuilder and it is "okay" to use it as long as you can control the version of gRPC being used (which generally excludes libraries from using experimental APIs). But we do want people to strongly consider if they are willing to trade guaranteed API stability for whatever feature they want from NettyServerBuilder.

**Comment by ejona86** [comment]
Fixed by #8758. As I mention in that PR:
> This is now much easier since Netty 4.1.70 which split the Java class files into a separate dependency.
